{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9b33049d",
   "metadata": {},
   "source": [
    "# COSMX COUNT QC REPORT FOR SLIDE {{ COSMX_SLIDE_NAME }}\n",
    "* **Notebook version:** v0.0.1\n",
    "* **Created by:** NIHR Imperial BRC Genomics Facility\n",
    "* **Maintained by:** NIHR Imperial BRC Genomics Facility\n",
    "* **Docker image path:** [Dockerfile](https://github.com/imperial-genomics-facility/igf-dockerfiles/tree/main/cosmx/Dockerfile_v1)\n",
    "* **Notebook code path:** [Templates](https://github.com/imperial-genomics-facility/igf-dockerfiles/tree/main/cosmx/)\n",
    "* **Created on:** {{ DATE_TAG }}\n",
    "* **Contact us:** [NIHR Imperial BRC Genomics Facility - Contact us](https://www.imperial.ac.uk/medicine/research-and-impact/facilities/genomics-facility/contact-us/)\n",
    "* **License:** Apache [License 2.0](https://github.com/imperial-genomics-facility/igf-dockerfiles/blob/main/LICENSE)\n",
    "\n",
    "* **Project name:** {{ COSMX_PROJECT_NAME }}\n",
    "* **COSMX slide name:** {{ COSMX_SLIDE_NAME }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b97c57fc-fff1-4c6a-a6fe-70edce596ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from typing import List, Tuple\n",
    "from IPython.display import HTML, Markdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abae485f-9a61-4f03-b61f-4f4d2b1a8383",
   "metadata": {},
   "outputs": [],
   "source": [
    "SLIDE_FLAT_FILE_DIR = \"{{ SLIDE_FLAT_FILE_DIR }}\"\n",
    "JSON_OUTPUT_DIR = \"{{ JSON_OUTPUT_DIR }}\"\n",
    "SLIDE_METADATA_JSON_FILE = \"{{ SLIDE_METADATA_JSON_FILE }}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67649e40-d14e-4e14-8071-afe2f35a355a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def get_assay_and_panel_info_from_slide_meatadata(slide_metadata_json: str, assay_type_key: str = 'assay_type', panel_name_key: str = 'panel_name') -> str:\n",
    "    json_data = {}\n",
    "    with open(slide_metadata_json, 'r') as fp:\n",
    "        json_data = json.load(fp)\n",
    "    assay_type = json_data.get(assay_type_key)\n",
    "    panel_name = json_data.get(panel_name_key)\n",
    "    if assay_type is None or panel_name is None:\n",
    "        raise KeyError(f\"Failed to get assay type or panel name from file {slide_metadata_json}\")\n",
    "    return assay_type, panel_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddb42e5-eea4-4167-82f4-9a2f59c0340b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def get_expr_mat_file_for_slide(slide_flat_file_dir: str) -> str:\n",
    "    slide_flat_file_path = Path(slide_flat_file_dir)\n",
    "    if not slide_flat_file_path.exists():\n",
    "        raise IOError(f\"FlatFiles dir {slide_flat_file_dir} not found\")\n",
    "    expr_mat_file = [\n",
    "        file_name for file_name in slide_flat_file_path.glob(\"*exprMat_file.csv.gz\")]\n",
    "    if len(expr_mat_file) == 0:\n",
    "        raise ValueError(f\"No exprMat_file.csv.gz found in {slide_flat_file_dir}\")\n",
    "    expr_mat_file = expr_mat_file[0]\n",
    "    return expr_mat_file.as_posix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0956ecc8-eadf-4395-aed8-c5cf10d2d3d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def get_rna_expr_counts(exprMat_file: str) -> Tuple[List[dict], List[dict]]:\n",
    "    try:\n",
    "        expr_df = \\\n",
    "            pd.read_csv(\n",
    "                exprMat_file,\n",
    "                compression=\"gzip\")\n",
    "        count_qc_data = []\n",
    "        ## get columns\n",
    "        sys_ctrl_cols = [c for c in expr_df if c.startswith(\"SystemControl\")]\n",
    "        neg_cols = [c for c in expr_df if c.startswith(\"Negative\")]\n",
    "        gene_cols = [c for c in expr_df if c!='fov' and c!='cell_ID' and c not in sys_ctrl_cols and c not in neg_cols]\n",
    "        ## count per fov\n",
    "        for fov_id, f_data in expr_df.groupby('fov'):\n",
    "            tmp_df = f_data.copy()\n",
    "            tmp_df['total_gene_count'] = tmp_df[gene_cols].sum(axis=1)\n",
    "            tmp_df['total_neg_count'] = tmp_df[neg_cols].sum(axis=1)\n",
    "            tmp_df['total_sys_count'] = tmp_df[sys_ctrl_cols].sum(axis=1)\n",
    "            min_neg_probe = tmp_df['total_neg_count'].mean()\n",
    "            mean_transcript_per_cell = tmp_df['total_gene_count'].mean()\n",
    "            percentile_90_transcript_per_cell = tmp_df['total_gene_count'].quantile(0.9)\n",
    "            percentile_10_transcript_per_cell = tmp_df['total_gene_count'].quantile(0.1)\n",
    "            total_cells = len(tmp_df.index)\n",
    "            non_empty_cells = len(tmp_df[tmp_df['total_gene_count']>0].index)\n",
    "            tmp_df['genes_per_cell'] = (tmp_df[gene_cols]>0).sum(axis=1)\n",
    "            mean_genes_per_cell = tmp_df['genes_per_cell'].mean()\n",
    "            count_qc_data.append({\n",
    "                'FOV': fov_id,\n",
    "                'Mean transcript per cell': f\"{mean_transcript_per_cell:.2f}\",\n",
    "                'Mean unique genes per cell': f\"{mean_genes_per_cell:.2f}\",\n",
    "                'Number of Non-empty cells': f\"{non_empty_cells}\",\n",
    "                'PCT of Non-empty cells': f\"{(non_empty_cells / total_cells):.2f}\",\n",
    "                '10th percentile transcript per cell': f\"{percentile_90_transcript_per_cell:.1f}\",\n",
    "                '90th percentile transcript per cell': f\"{percentile_10_transcript_per_cell:.1f}\",\n",
    "                'Mean Negprobe counts per cell': f\"{min_neg_probe:.3f}\"})\n",
    "            df = pd.DataFrame(count_qc_data)\n",
    "            df.rename(\n",
    "                columns={\n",
    "                    'FOV': \"fov_id\",\n",
    "                    'Mean transcript per cell': 'mean_transcript_per_cell',\n",
    "                    'Mean unique genes per cell': 'mean_unique_genes_per_cell',\n",
    "                    'Number of Non-empty cells': 'number_non_empty_cells',\n",
    "                    'PCT of Non-empty cells': 'pct_non_empty_cells',\n",
    "                    '10th percentile transcript per cell': 'percentile_10_transcript_per_cell',\n",
    "                    '90th percentile transcript per cell': 'percentile_90_transcript_per_cell',\n",
    "                    'Mean Negprobe counts per cell': 'mean_negprobe_counts_per_cell'},\n",
    "                inplace=True)\n",
    "            df = \\\n",
    "                df.astype({\n",
    "                    \"fov_id\": int,\n",
    "                    'mean_transcript_per_cell': float,\n",
    "                    'mean_unique_genes_per_cell': float,\n",
    "                    'number_non_empty_cells': int,\n",
    "                    'pct_non_empty_cells': float,\n",
    "                    'percentile_10_transcript_per_cell': float,\n",
    "                    'percentile_90_transcript_per_cell': float,\n",
    "                    'mean_negprobe_counts_per_cell': float})\n",
    "            count_qc_data_for_db = \\\n",
    "                df.todict(orient=\"records\")\n",
    "        return count_qc_data, count_qc_data_for_db\n",
    "    except Exception as e:\n",
    "        raise ValueError(f\"Failed to get RNA QC for file {exprMat_file}, error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "156f3155-c7c3-4462-841a-830d24aba068",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def get_protein_expr_counts(exprMat_file: str) -> Tuple[List[dict], List[dict]]:\n",
    "    try:\n",
    "        expr_df = \\\n",
    "            pd.read_csv(\n",
    "                exprMat_file,\n",
    "                compression=\"gzip\")\n",
    "        igG_cols = [c for c in expr_df.columns if 'IgG' in c]\n",
    "        prot_cols = [p for p in expr_df.columns if p not in igG_cols and p != \"fov\" and p != \"cell_ID\"]\n",
    "        count_qc_data = []\n",
    "        ## count per fov\n",
    "        for fov_id, f_data in expr_df.groupby('fov'):\n",
    "            tmp_df = f_data.copy()\n",
    "            tmp_df['total_counts'] = tmp_df[prot_cols].sum(axis=1)\n",
    "            tmp_df['total_prot_counts'] = tmp_df[prot_cols+igG_cols].sum(axis=1)\n",
    "            tmp_df['genes_per_cell'] = (tmp_df[prot_cols]>0).sum(axis=1)\n",
    "            count_qc_data.append({\n",
    "                'FOV': fov_id,\n",
    "                'Mean Fluorescence Intensity (MFI)': f\"{tmp_df['total_counts'].mean():.0f}\",\n",
    "                'Mean unique genes per cell': f\"{tmp_df['genes_per_cell'].mean():.0f}\",\n",
    "                'Number of Non-empty cells': len(tmp_df[tmp_df['total_counts']>0].index),\n",
    "                'PCT of Non-empty cells': f\"{len(tmp_df[tmp_df['total_counts']>0].index) / len(tmp_df.index):.2f}\",\n",
    "                '10th percentile of Fluorescence Intensity': f\"{tmp_df['total_counts'].quantile(0.1):.2f}\",\n",
    "                '90th percentile of Fluorescence Intensity': f\"{tmp_df['total_counts'].quantile(0.9):.2f}\",\n",
    "                'Fluorescence Intensity\tMean IgG Control Intensity': f\"{tmp_df[igG_cols].sum(axis=1).mean():.3f}\"\n",
    "            })\n",
    "        df = pd.DataFrame(count_qc_data)\n",
    "            df.rename(\n",
    "                columns={\n",
    "                    'FOV': \"fov_id\",\n",
    "                    'Mean Fluorescence Intensity (MFI)': 'mean_fluorescence_intensity',\n",
    "                    'Mean unique genes per cell': 'mean_unique_genes_per_cell',\n",
    "                    'Number of Non-empty cells': 'number_non_empty_cells',\n",
    "                    'PCT of Non-empty cells': 'pct_non_empty_cells',\n",
    "                    '10th percentile of Fluorescence Intensity': 'percentile_10_fluorescence_intensity',\n",
    "                    '90th percentile of Fluorescence Intensity': 'percentile_90_fluorescence_intensity',\n",
    "                    'Fluorescence Intensity\tMean IgG Control Intensity': 'fluorescence_intensity_mean_igg_control_intensity'\n",
    "                    },\n",
    "                inplace=True)\n",
    "            df = \\\n",
    "                df.astype({\n",
    "                    \"fov_id\": int,\n",
    "                    \"mean_fluorescence_intensity\": int,\n",
    "                    \"mean_unique_genes_per_cell\": int,\n",
    "                    \"number_non_empty_cells\": int,\n",
    "                    \"pct_non_empty_cells\": float,\n",
    "                    \"percentile_10_fluorescence_intensity\": float,\n",
    "                    \"percentile_90_fluorescence_intensity\": float,\n",
    "                    \"fluorescence_intensity_mean_igg_control_intensity\": float}})\n",
    "            count_qc_data_for_db = \\\n",
    "                df.todict(orient=\"records\")\n",
    "        return count_qc_data, count_qc_data_for_db\n",
    "    except Exception as e:\n",
    "        raise ValueError(f\"Failed to get Protein QC for file {exprMat_file}, error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b3110e3-a398-4fd7-ab8b-03b4b4363a60",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def generate_count_qc_table(\n",
    "    slide_flat_file_dir: str,\n",
    "    slide_metadata_json_file: str,\n",
    "    json_output_file: str ) -> Tuple[List[dict], str]:\n",
    "    assay_type, panel_name = \\\n",
    "        get_assay_and_panel_info_from_slide_meatadata(\n",
    "            slide_metadata_json_file)\n",
    "    expr_mat_file = \\\n",
    "        get_expr_mat_file_for_slide(slide_flat_file_dir)\n",
    "    count_qc_data = list({})\n",
    "    if assay_type.upper() == 'RNA':\n",
    "        count_qc_data, count_qc_data_for_db = \\\n",
    "            get_rna_expr_counts(exprMat_file=expr_mat_file)\n",
    "    elif assay_type.upper() == 'PROTEIN':\n",
    "        count_qc_data, count_qc_data_for_db = \\\n",
    "            get_protein_expr_counts(exprMat_file=expr_mat_file)\n",
    "    else:\n",
    "        raise ValueError(f\"New assay type {assay_type} found in COSMX metadata flat file {slide_metadata_json_file}\")\n",
    "    with open(json_output_file, 'w') as fp:\n",
    "        json.dump(count_qc_data_for_db, fp)\n",
    "    return count_qc_data, assay_type, panel_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d73a682f-067a-4518-8216-537d1ec3a2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_qc_data, assay_type, panel_name = \\\n",
    "    generate_count_qc_table(\n",
    "        slide_flat_file_dir=SLIDE_FLAT_FILE_DIR,\n",
    "        slide_metadata_json_file=SLIDE_METADATA_JSON_FILE,\n",
    "        json_output_file=JSON_OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8778e30d-e08f-47bf-85e3-49eea0d4bf8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Markdown(f\"\"\"* **Assay type:** {assay_type}\n",
    "* **Panel name:** {panel_name}\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5da498b9-3c70-4609-93e4-9107ab98ed6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML(pd.DataFrame(count_qc_data).to_html(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4d2356c-0532-462e-b32d-3014bdc59564",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "airflow_2.10.4",
   "language": "python",
   "name": "airflow_2.10.4"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
